File: android/mdi-android/src/main/kotlin/app/illusions/mdi/Mdi.kt
package app.illusions.mdi

import app.illusions.mdi.internal.MdiNative
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

/** MDI syntax version supported by this Android binding. */
public const val MDI_SPEC_VERSION: String = "2.0"

/** Version of the Rust-owned document IR understood by this binding. */
public const val MDI_IR_VERSION: String = "1.0"

/** Complete Android interface to the Rust-authoritative MDI implementation. */
public object Mdi {
    /** Parses complete MDI source; Kotlin does not tokenize or repair syntax. */
    @JvmStatic
    public fun parse(source: String): MdiParseResult {
        val result = MdiJson.decode(MdiBridgeHolder.bridge.parseJson(source))
        require(result.irVersion == MDI_IR_VERSION) {
            "Unsupported MDI IR version: ${result.irVersion}"
        }
        return result
    }

    /** Renders standalone HTML from the Rust-owned document IR. */
    @JvmStatic public fun renderHtml(source: String): String = MdiBridgeHolder.bridge.renderHtml(source)

    /** Canonically serializes source through the Rust parser. */
    @JvmStatic public fun serializeMdi(source: String): String = MdiBridgeHolder.bridge.serializeMdi(source)

    /** Renders deterministic plain text through Rust. */
    @JvmStatic public fun renderText(source: String): String = MdiBridgeHolder.bridge.renderText(source)

    /** Renders a named publication-text convention through Rust. */
    @JvmStatic
    public fun renderTextFormat(source: String, format: MdiTextFormat, indentPrefix: String = ""): String =
        MdiBridgeHolder.bridge.renderTextFormat(source, format.wireName, indentPrefix)

    /** Builds an EPUB archive. Persist the returned bytes with Android's Storage Access Framework. */
    @JvmStatic public fun renderEpub(source: String): ByteArray = MdiBridgeHolder.bridge.renderEpub(source)

    /** Builds a DOCX archive. Persist the returned bytes with Android's Storage Access Framework. */
    @JvmStatic public fun renderDocx(source: String): ByteArray = MdiBridgeHolder.bridge.renderDocx(source)
}

/** Text export conventions implemented by mdi-core. */
public enum class MdiTextFormat(internal val wireName: String) {
    Plain("txt"), Ruby("txt-ruby"), Narou("narou"), Kakuyomu("kakuyomu"), Aozora("aozora"),
}

@Serializable
public data class MdiParseResult(
    val irVersion: String,
    val syntaxVersion: String,
    val capabilities: MdiParserCapabilities,
    /** Rust's tagged, extensible document IR. Do not infer MDI grammar from it. */
    val document: JsonObject,
    val diagnostics: List<MdiDiagnostic>,
)

@Serializable
public data class MdiParserCapabilities(
    val mdi: Boolean,
    val commonMark: Boolean,
    val gfm: Boolean,
    val frontMatter: Boolean,
    val sourceSpans: Boolean,
)

@Serializable
public data class MdiDiagnostic(
    val severity: String,
    val code: String,
    val message: String,
    val span: MdiSourceSpan? = null,
)

/** Half-open UTF-8 byte range, not a Kotlin String character range. */
@Serializable
public data class MdiSourceSpan(
    val startByte: Int,
    val endByte: Int,
)

internal object MdiJson {
    private val json = Json { ignoreUnknownKeys = false }

    fun decode(value: String): MdiParseResult = json.decodeFromString(value)
}

/** Host boundary kept separate so unit tests can cover the public API without JNI. */
internal interface MdiBridge {
    fun parseJson(source: String): String
    fun renderHtml(source: String): String
    fun serializeMdi(source: String): String
    fun renderText(source: String): String
    fun renderTextFormat(source: String, format: String, indentPrefix: String): String
    fun renderEpub(source: String): ByteArray
    fun renderDocx(source: String): ByteArray
}

internal object NativeMdiBridge : MdiBridge {
    override fun parseJson(source: String): String = MdiNative.parseJson(source)
    override fun renderHtml(source: String): String = MdiNative.renderHtml(source)
    override fun serializeMdi(source: String): String = MdiNative.serializeMdi(source)
    override fun renderText(source: String): String = MdiNative.renderText(source)
    override fun renderTextFormat(source: String, format: String, indentPrefix: String): String =
        MdiNative.renderTextFormat(source, format, indentPrefix)
    override fun renderEpub(source: String): ByteArray = MdiNative.renderEpub(source)
    override fun renderDocx(source: String): ByteArray = MdiNative.renderDocx(source)
}

internal object MdiBridgeHolder {
    var bridge: MdiBridge = NativeMdiBridge
}

File: swift/Sources/MDI/MDI.swift
import Foundation
import MDICore

/// MDI's supported language-specification version.
public let mdiSpecVersion = "2.0"

/// The version of the Rust-owned JSON document contract.
public let mdiIRVersion = "1.0"

public enum MDIError: Error, Equatable, LocalizedError, Sendable {
    case core(String)
    case invalidWireFormat(String)

    public var errorDescription: String? {
        switch self {
        case let .core(message), let .invalidWireFormat(message):
            return message
        }
    }
}

public struct MDISourceSpan: Codable, Equatable, Sendable {
    public let startByte: UInt32
    public let endByte: UInt32
}

public struct MDIParserCapabilities: Codable, Equatable, Sendable {
    public let mdi: Bool
    public let commonMark: Bool
    public let gfm: Bool
    public let frontMatter: Bool
    public let sourceSpans: Bool
}

public enum MDIDiagnosticSeverity: String, Codable, Sendable { case warning, error }

public struct MDIDiagnostic: Codable, Equatable, Sendable {
    public let severity: MDIDiagnosticSeverity
    public let code: String
    public let message: String
    public let span: MDISourceSpan?
}

/// Lossless representation of a value in the versioned MDI document IR.
public indirect enum MDIJSONValue: Codable, Equatable, Sendable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([MDIJSONValue])
    case object([String: MDIJSONValue])

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() { self = .null }
        else if let value = try? container.decode(Bool.self) { self = .bool(value) }
        else if let value = try? container.decode(Double.self) { self = .number(value) }
        else if let value = try? container.decode(String.self) { self = .string(value) }
        else if let value = try? container.decode([MDIJSONValue].self) { self = .array(value) }
        else { self = .object(try container.decode([String: MDIJSONValue].self)) }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null: try container.encodeNil()
        case let .bool(value): try container.encode(value)
        case let .number(value): try container.encode(value)
        case let .string(value): try container.encode(value)
        case let .array(value): try container.encode(value)
        case let .object(value): try container.encode(value)
        }
    }
}

public struct MDIParseResult: Codable, Equatable, Sendable {
    public let irVersion: String
    public let syntaxVersion: String
    public let capabilities: MDIParserCapabilities
    public let document: MDIJSONValue
    public let diagnostics: [MDIDiagnostic]
}

/// Publication text conventions implemented by the Rust core.
public enum MDITextFormat: String, CaseIterable, Codable, Sendable {
    case plain = "txt"
    case ruby = "txt-ruby"
    case narou
    case kakuyomu
    case aozora
    case note
}

/// Swift interface to the Rust-owned MDI parser and renderers.
public enum MDI {
    public static func parse(_ source: String) throws -> MDIParseResult {
        try parseResult(from: call(source, operation: mdi_parse_json))
    }

    static func parseResult(from data: Data) throws -> MDIParseResult {
        let result: MDIParseResult
        do { result = try JSONDecoder().decode(MDIParseResult.self, from: data) }
        catch { throw MDIError.invalidWireFormat("MDI core returned invalid parse JSON: \(error.localizedDescription)") }
        guard result.irVersion == mdiIRVersion else {
            throw MDIError.invalidWireFormat("Unsupported MDI IR version: \(result.irVersion)")
        }
        return result
    }

    public static func renderHTML(_ source: String) throws -> String { try stringCall(source, operation: mdi_render_html) }
    public static func serialize(_ source: String) throws -> String { try stringCall(source, operation: mdi_serialize_mdi) }
    public static func renderText(_ source: String) throws -> String { try stringCall(source, operation: mdi_render_text) }
    public static func renderTextFormat(
        _ source: String,
        format: MDITextFormat,
        indentPrefix: String = ""
    ) throws -> String {
        let sourceBytes = Array(source.utf8)
        let formatBytes = Array(format.rawValue.utf8)
        let indentBytes = Array(indentPrefix.utf8)
        let result = sourceBytes.withUnsafeBufferPointer { sourceBuffer in
            formatBytes.withUnsafeBufferPointer { formatBuffer in
                indentBytes.withUnsafeBufferPointer { indentBuffer in
                    mdi_render_text_format(
                        sourceBuffer.baseAddress,
                        sourceBuffer.count,
                        formatBuffer.baseAddress,
                        formatBuffer.count,
                        indentBuffer.baseAddress,
                        indentBuffer.count
                    )
                }
            }
        }
        return try string(from: data(from: result))
    }
    public static func renderEPUB(_ source: String) throws -> Data { try call(source, operation: mdi_render_epub) }
    public static func renderDOCX(_ source: String) throws -> Data { try call(source, operation: mdi_render_docx) }

    static func string(from data: Data) throws -> String {
        guard let value = String(data: data, encoding: .utf8) else {
            throw MDIError.invalidWireFormat("MDI core returned non-UTF-8 string data")
        }
        return value
    }

    static func coreErrorMessage(from data: Data) -> String {
        String(data: data, encoding: .utf8) ?? "Unknown MDI core error"
    }

    private static func stringCall(_ source: String, operation: MDIFFIOperation) throws -> String {
        try string(from: call(source, operation: operation))
    }

    static func call(_ source: String, operation: MDIFFIOperation) throws -> Data {
        try call(bytes: Array(source.utf8), operation: operation)
    }

    static func call(bytes: [UInt8], operation: MDIFFIOperation) throws -> Data {
        let result = bytes.withUnsafeBufferPointer { operation($0.baseAddress, $0.count) }
        return try data(from: result)
    }

    static func data(from result: mdi_ffi_result) throws -> Data {
        defer { mdi_free_buffer(result.value); mdi_free_buffer(result.error) }
        if result.error.len > 0 {
            let message = coreErrorMessage(from: Data(bytes: result.error.data!, count: result.error.len))
            throw MDIError.core(message)
        }
        guard result.value.len == 0 || result.value.data != nil else {
            throw MDIError.invalidWireFormat("MDI core returned an invalid buffer")
        }
        return result.value.len == 0 ? Data() : Data(bytes: result.value.data!, count: result.value.len)
    }
}

typealias MDIFFIOperation = @convention(c) (UnsafePointer<UInt8>?, Int) -> mdi_ffi_result
